core/blockchain/blockchain_gateway.py

```diff
@@ -0,0 +1,355 @@
+        """
+        Polls blockchain for node ID in decentralized_learning() method
+        signature decentralized_learning(...<node_ids>...) method signature
+        will be the key on the blockchain; listener should look for this, and
+        if the method signature contains its node id, it will trigger a
+        callback
+        """
+        def filter(tx):
+            logging.info("tx: {}".format(tx))
+            return tx.get('key') == tx.get('content')
+        return self.filter_set(filter,
+                                self.handle_decentralized_learning_trainer)
+
+    def broadcast_new_weights(self, tx: dict) -> str:
+        """
+        broadcast_new_weights() method with all relevant parameters
+        should populate the key of new_weights with all of the nodes for which
+        these new weights are relevant. value should be IPFS hash.
+        """
+        content = tx.get('content', None)
+        key = self.client.add_json(content['weights'])
+        tx_receipt = self.setter(key, content['gradient'])
+        return tx_receipt
+
+    def listen_new_weights(self):
+        """
+        Polls blockchain for node ID in new_weights() method signature
+        new_weights(...<node_ids>...) method signature will be the key on the
+        blockchain; listener should look for this, and if the method signature
+        contains its node id, it will trigger a callback
+        """
+        logging.info("I'm listening for new weights!")
+        def weights_filter(tx):
+            logging.info("filtering for new weights: {}".format(tx))
+            return tx.get('key') != tx.get('content')
+        self.filter_set(weights_filter, self.handle_new_weights)
+
+    def listen_terminate(self):
+        """
+        Polls blockchain to see whether to terminate
+        """
+        def filter(tx):
+            logging.info("tx: {}".format(tx))
+            return tx.get('content') is None
+        self.filter_set(filter, self.handle_terminate)
+
+    def inform(self, event_type, payload):
+        """
+        Method called by other modules to inform the Listener about
+        events that are going on in the service.
+        These payloads are relayed to the blockchain (right now the only
+        one, in the future, the one corresponding to the session_id passed),
+        based on some internal logic of the Listener.
+        For example: A runner could inform the Communication Manager that the
+        node is done training a particular model, to which the Optimizer could
+        decide it's time to communicate the new weights to the network.
+        If the Optimizer says yes, then the Communication Manager should
+        relay this info to the Listener, and the Listener uploads weights.
+        """
+        self._parse_and_run_callback(event_type, payload)
+
+    def _parse_and_run_callback(self, event_type, payload):
+        """
+        Parses an actionable_event and runs the
```

Comment:
Shouldn't it be "Parses a message_event..."?

Reply:
yep i copy-pasted whoops will update docstring

Reply:
@nzoghb after you commit, I'll update this to use the callback_handler convention that's going to be on master